metrics-core/src/lib.rs

```diff
@@ -59,3 +60,28 @@ pub trait MetricsRecorder {
     /// of the underlying observed values, and callers will need to process them accordingly.
     fn record_histogram<K: AsRef<str>>(&mut self, key: K, values: &[u64]);
 }
+
+/// A value that holds a point-in-time view of collected metrics.
+pub trait MetricsSnapshot {
+    /// Records the snapshot to the given recorder.
+    fn record<R: MetricsRecorder>(&self, recorder: &mut R);
+}
+
+/// A value that can provide on-demand snapshots.
+pub trait SnapshotProvider {
+    type Snapshot;
```

Comment:
shouldn't this be `type Snapshot: MetricSnapshot`?

Reply:
Good call! This lets me drop all those constraints on the exporter. :D